pkg/deploy/rbac/gateway_permissions.go

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package rbac
+
+import (
+	orgv1 "github.com/eclipse-che/che-operator/api/v1"
+	"github.com/eclipse-che/che-operator/pkg/deploy"
+	"github.com/eclipse-che/che-operator/pkg/deploy/gateway"
+	"github.com/eclipse-che/che-operator/pkg/util"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	CheGatewayClusterPermissionsFinalizerName = "cheGateway.clusterpermissions.finalizers.che.eclipse.org"
+)
+
+type GatewayPermissionsReconciler struct {
+	deploy.Reconcilable
+}
+
+func NewGatewayPermissionsReconciler() *GatewayPermissionsReconciler {
+	return &GatewayPermissionsReconciler{}
+}
+
+func (gp *GatewayPermissionsReconciler) Reconcile(ctx *deploy.DeployContext) (reconcile.Result, bool, error) {
+	if util.IsOpenShift && ctx.CheCluster.IsNativeUserModeEnabled() {
+		name := gp.gatewayPermissionsName(ctx.CheCluster)
+		if done, err := deploy.SyncClusterRoleToCluster(ctx, name, gp.getGatewayClusterRoleRules()); !done {
+			return reconcile.Result{Requeue: true}, false, err
+		}
+
+		if done, err := deploy.SyncClusterRoleBindingToCluster(ctx, name, gateway.GatewayServiceName, name); !done {
+			return reconcile.Result{Requeue: true}, false, err
+		}
+
+		if err := deploy.AppendFinalizer(ctx, CheGatewayClusterPermissionsFinalizerName); err != nil {
+			return reconcile.Result{Requeue: true}, false, err
+		}
+	} else {
+		if done, err := gp.deleteGatewayPermissions(ctx); !done {
+			return reconcile.Result{Requeue: true}, false, err
+		}
+	}
+
+	return reconcile.Result{}, true, nil
+}
+
+func (gp *GatewayPermissionsReconciler) Finalize(ctx *deploy.DeployContext) error {
+	_, err := gp.deleteGatewayPermissions(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (gp *GatewayPermissionsReconciler) deleteGatewayPermissions(deployContext *deploy.DeployContext) (bool, error) {
+	name := gp.gatewayPermissionsName(deployContext.CheCluster)
+	if done, err := deploy.Delete(deployContext, types.NamespacedName{Name: name}, &rbacv1.ClusterRole{}); !done {
```

Comment:
Shouldn't we delete `RoleBinding` first?

Reply:
I think yes